design docs and a diagram for vehicle dynamics

adamdbrw opened the pull request:
Fixes #144

Signed-off-by: Adam Dabrowski <adam.dabrowski@robotec.ai>

adamdbrw committed:
design docs and a diagram for vehicle dynamics

Fixes #144

Signed-off-by: Adam Dabrowski <adam.dabrowski@robotec.ai>

piotr-zyskowski-rai commented on docs/design/vehicle_dynamics/analysis.md:
```diff
@@ -0,0 +1,71 @@
+
+## Types of robot mobility and the scope of analysis
+
+This analysis only includes **wheeled robots**.
```
To me seems more logical to put this disclaimer below the list of types of robot mobility.

It might also be handy to emphasize this message a bit more. For example using table header

```
| This analysis only includes wheeled robots |
|--------------------------------------------|
```
Which looks good both in the raw text file as well as generated

piotr-zyskowski-rai commented on docs/design/vehicle_dynamics/analysis.md:
```diff
@@ -0,0 +1,71 @@
+
+### Car-like steering (e.g., Ackerman Steering or Side Pivot Steering)
+
+Car steering avoids side slip on wheels when moving on a curve, which is especially important for stability at high speeds.
```
Less slippage also makes the odometry sensors of this steering type more reliable.

piotr-zyskowski-rai commented on docs/design/vehicle_dynamics/analysis.md:
```diff
@@ -0,0 +1,71 @@
+
+### Differential drive
+
+Two independent wheels, steering is achieved through different rates of rotation. Typically, a third unpowered wheel for balance is added.
```
Also, heavy machinery with tracks is an example of differential drive. Many experimental robots also have four-wheeled differential drive.

j-rivero replied:
This also caught my attention, four-wheeled vehicles is what comes to my mind when we speak about differential drives.

piotr-zyskowski-rai commented on docs/design/vehicle_dynamics/analysis.md:
```diff
@@ -0,0 +1,71 @@
+# Vehicle Physics in O3DE - analysis
```
General remark: Some pictures might enrich this design doc substantially

j-rivero replied:
+1 a [quick search for creative common images on Google](https://www.google.com/search?q=vehicle%20type%20of%20wheels%20robotics&tbm=isch&hl=en&tbs=itp:lineart&client=ubuntu&hs=kYK&sa=X&ved=0CAMQpwVqFwoTCPCAuPvHgPoCFQAAAAAdAAAAABAD&biw=1920&bih=968) could help to find quick examples

piotr-zyskowski-rai commented on docs/design/vehicle_dynamics/design.md:
```diff
@@ -0,0 +1,25 @@
+
+### Subsections
+
+- [Analysis](analisys.md)
```
```suggestion
- [Analysis](analysis.md)
```

piotr-zyskowski-rai commented on docs/design/vehicle_dynamics/design.md:
```diff
@@ -0,0 +1,25 @@
+
+### Note
+
+In the first iteration, only support for car-like 4-wheeled robots will be targeted.
```
```suggestion
| In the first iteration, only support for car-like 4-wheeled robots will be targeted. |
|--------------------------------------------------------------------------------------|
```

piotr-zyskowski-rai commented on docs/design/vehicle_dynamics/features.md:
```diff
@@ -0,0 +1,20 @@
+| Feature | Description | Importance |
+| ------- | ----------- | ---------- |
+| Modular and configurable design | <ul><li> Highly customizable – can support variety of robots/vehicles </li><li> Modular – easy to replace implementations of parts with alternatives </li><li> Good defaults – less steep learning curve </li></ul> | Crucial |
+| Vehicle configuration | Number and configuration of wheels, axles, etc.  | Crucial |
```
Inconsistent style with other table rows. I know there is a single item here but I would add a bullet point here as well
```suggestion
| Vehicle configuration | <ul><li> Number and configuration of wheels, axles, etc. </li></ul> | Crucial |
```

piotr-zyskowski-rai commented on docs/design/vehicle_dynamics/features.md:
```diff
@@ -0,0 +1,20 @@
+| Path following | <ul><li> Follow a given path with a configurable speed limit </li><li> Adjust speed when approaching curves </li><li> Optional: stop when something is in the way (based on ground truth). This could be useful for testing.</li></ul>  | Important |
+| Engine | <ul><li> Power curve, losses, limits </li></ul> | Optional |
+| Transmission | <ul><li> Automatic, Manual, etc. </li><li> Gears, gear ratios, shifting behavior </li></ul> | Optional |
+| Aerodynamics |  | Optional |
```
Intentionally empty?

adamdbrw replied:
Yes, but added a point there

pijaro commented on docs/design/vehicle_dynamics/analysis.md:
```diff
@@ -0,0 +1,71 @@
+- **Castor wheels** (and balls) - passive wheels used for balance. 
+- **Car-like wheels** – connected through axles, with additional steering elements, with tires. 
+
+## Selected robot steering types
```
I would love to see more explanations of how the different types are implemented (o3de joints and rigid bodies or PhysX vehicle framework?).

piotr-zyskowski-rai reviewed (commented):
One link typo and several notes to consider. Overall nice doc :+1:

j-rivero commented on docs/design/vehicle_dynamics/features.md:
```diff
@@ -0,0 +1,20 @@
+| Vehicle configuration | Number and configuration of wheels, axles, etc.  | Crucial |
+| Core dynamics | <ul><li> Realistically transfer inputs to robot/vehicle behavior</li><li>Multiple steps per simulation step (sub-step solver)</li><li>Considering vehicle parameters such as mass and inertia</li><li>Considering engine/transmission parameters, such as max torque, max angular velocity</li><li>Selection of drive wheels (minimum: front and/or rear)</li></ul> | Crucial |
+| Wheels and suspension | <ul><li>Handle ground detection</li><li>Apply friction model (speed dependent)</li><li>Use wheel 3D geometry for collision</li><li> Adjust camber, caster, rim offset </li><li> Handle wheel visuals (rotate, turn)</li><li> Handle spring and damper, adjustable curves </li><li> Type of steering (differential or car-like)</li><li>Selection of steering wheels </li></ul> | Cruc ...
+| ROS 2 Control | <ul><li> Setting a topic, QoS, handling subscriptions</li><li>Selecting message type for control – AckermanDrive or Twist</li><li>Applying control to the robot/vehicle. Stopping on no input. </li></ul>| Crucial |
```
For ROS2 the most-common way of controlling vehicles is by using the `/cmd/vel` topic. Unfortunately the interface has not been standardized (only mention in [turtlebot REP](https://ros.org/reps/rep-0119.html)). The topic usually receives a `geometry_msgs/Twist` although[ there are discussion](https://discourse.ros.org/t/is-twist-still-a-good-velocity-command-interface/13218/4) to extend it to something more detailed. Might be a good idea to mention part of this at least.


gazebo_ros_pkgs (for ROS1) [diffdrive plugin](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/noetic-devel/gazebo_plugins/src/gazebo_ros_diff_drive.cpp#L92-L103) has some parameters that could help to understand what needs to be used in ROS to have the simulation working. Same for ROS" [ackermann plugin](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/ros2/gazebo_plugins/src/gazebo_ros_ackermann_drive.cpp#L46-L192) and [diffdrive](https://github.com/ros-simulation/gazebo_ros_pkgs/blob/ros2/gazebo_plugins/src/gazebo_ros_diff_drive.cpp#L88-L224).

j-rivero commented on docs/design/vehicle_dynamics/architecture.md:
```diff
@@ -0,0 +1,5 @@
+# Architecture
+
+## Diagram
+
```
For reference and helping to be complete in the initial architecture (I don't see major problems) these are the parameters and communications used in gz-sim: [diffdrive](https://github.com/gazebosim/gz-sim/blob/ign-gazebo6/src/systems/diff_drive/DiffDrive.hh#L35-L131) and [ackermann](https://github.com/gazebosim/gz-sim/blob/ign-gazebo6/src/systems/ackermann_steering/AckermannSteering.hh.)

adamdbrw replied:
This is useful, thank you!

j-rivero reviewed (commented):
Not my expertise field, overall looks good to me. I just added some mentions to my experience on Gazebo/Gz/ROS2 that can help the design to be more complete or fix a potential missing feature.

lgleim commented:
Have you considered the possibility to integrate with the [ros2_control](https://control.ros.org/master/index.html) controller implementations similar to their [implementation for Gazebo](https://github.com/ros-controls/gz_ros2_control) instead of implementing differential drive / ackermann etc. from scratch? This would come with a few advantages such as a larger community implementing controllers in the future and supporting MoveIt out of the box.

adamdbrw commented:
@lgleim this is a good point and you are right that there would be value from it. It is also quite possible since we can use any package within the Gem with ease. 

I was aware of recent developments in ros2_control, but I believe there is no Ackermann controller implemented in ros2_control yet. We are focusing on car-like vehicle for the second milestone, but we also would like the design to accommodate variety of implementations. I would still likely use at least a PID controller implementation from ros2_control. Our first implementation would likely be quite simple and inspire a second phase of design to follow after the second milestone.

I will update the design document with all the feedback as well as my own changes coming from the fact that I am already implementing it and put some more thought into it.

Integrating with ros2_control and MoveIt is also planned within the Manipulator tasks.

lgleim commented:
Thanks for the quick response @adamdbrw. There is indeed no Ackermann controller, there is however a [tricycle controller](https://github.com/ros-controls/ros2_controllers/tree/master/tricycle_controller) which is not the same but [quite similar](https://www.quora.com/Which-steering-mechanism-should-I-use-if-I-want-to-make-a-tricycle-with-tadpole-design). 

Great to hear that integrating with ros2_control and MoveIt is also planned!

adamdbrw commented:
@lgleim I looked a bit more into it and I see a bit of a challenge. There are some assumptions about control loops and the controller is integrated with publishing and parameters. In short, I would like to take some parts of it, but less so the entire thing. For example, simulators have their control loops. Another example is that ideally, parameters could be handled in the Editor mode and synchronized between UI and parameters file, while publishing should only happen in the Play mode. The controller in its current form seems a bit difficult to integrate.

It could run much like a plugin though - and this is something we should consider. Would you like to take a part in the design process? We could also make contributions to ros2_control since I suppose it is has been only used with one simulator so far - Gazebo. I would propose to do this after milestone 2 though. Perhaps there will be an Ackermann controller implementation already, ROSCon speeds some things up ;)

Also, a question is what should really run in the simulation? Would the same controller also run on the real robot? If so, then it should be outside of the simulation in the robotic stack package which is launched side-by-side.

michalpelka commented on docs/design/vehicle_dynamics/analysis.md:
```diff
@@ -0,0 +1,71 @@
+    - …
+  - Crawling robots (snake-like)
+  - Climbing robots  
+  - ... 
```
Please consider adding tracked robot (like e.g. iRobot packbot)

michalpelka commented on docs/design/vehicle_dynamics/analysis.md:
```diff
@@ -0,0 +1,71 @@
+
+## Selected robot steering types
+
+### Differential drive
```
Maybe change to skid-steering.

adamdbrw replied:
Added skid-steering to the title, but kept the original text as well

lgleim commented:
@adamdbrw That are exactly the questions we have been asking ourselves as well. ros2_control goes as far as extending the URDF model with custom annotations which require importer support. We are generally leaning towards making it possible to reuse controllers that would be running on real robots but also supporting simplifications (such as sticking to direct  twist control) to simulate higher-level tasks (e.g. development work on path-planning) faster. As such, having dedicated implementations (or reusing lower-level ros2_control implementation) for simple controllers within this gem sounds like the right first step. That being said, we would definitely like to take part in the design process.

michalpelka commented on docs/design/vehicle_dynamics/features.md:
```diff
@@ -0,0 +1,20 @@
+| Aerodynamics |  | Optional |
+| Visuals and sounds | <ul><li> Lights </li><li> Engine and other sounds </li><li> Skid marks  </li><li>  Particle effects, exhaust smoke </li></ul> | Optional |
+| Transmission | <ul><li> Power distribution between wheels </li></ul> | Optional |
+| Axles | <ul><li> Brake force distribution between wheels </li></ul>| Optional |
```
add differential mechanism and its configuration.

michalpelka reviewed (commented):
Loose comments from my side.

adamdbrw committed:
Updated diagram, added corrections, sections and pictures as for review comments

Signed-off-by: Adam Dabrowski <adam.dabrowski@robotec.ai>

adamdbrw commented:
Review comments have been applied:
- @piotr-zyskowski-rai , @pijaro , @michalpelka: all review comments applied
- @pijaro, @lgleim: I added new sections in Design Considerations according to your comments
- @j-rivero I added some context as you suggested and added a couple of pictures (along source / license info)
- Additionally, I updated / improved the architecture diagram. No revolutonary changes there, but some of note: take a look.

@lgleim 
> That being said, we would definitely like to take part in the design process

Wonderful! I am looking forward to it!

adamdbrw commented:
@spham-amzn I also invite you to take a look and perhaps invite someone you find suitable
